### PR TITLE
MULE-13633 Upgrade to Project Reactor 3.1.0

### PR DESCRIPTION
**Pull request by dfeist**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/ReactorProcessingStrategyTestCase.java [hunk 1]
```diff
@@ -200,7 +200,7 @@ public void concurrentStream() throws Exception {
   @Description("If CPU LITE pool is busy OVERLOAD error is thrown")
   public void cpuLightRejectedExecution() throws Exception {
     flow = flowBuilder.get().processors(blockingProcessor)
-        .processingStrategyFactory((context, prefix) -> new ReactorProcessingStrategy(() -> new RejectingScheduler(blocking)))
+        .processingStrategyFactory((context, prefix) -> new ReactorProcessingStrategy(() -> new RejectingScheduler(cpuLight)))
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/ReactorProcessingStrategyTestCase.java]: why this change?

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/AbstractProcessingStrategy.java [hunk 2]
```diff
@@ -57,32 +55,39 @@ protected ExecutorService decorateScheduler(Scheduler scheduler) {
...
-      blockingSink.accept(event);
+      fluxSink.next(event);
     }
 
     @Override
     public boolean emit(CoreEvent event) {
       onEventConsumer.accept(event);
-      return blockingSink.emit(event) != BACKPRESSURED;
+      synchronized (fluxSink) {
+        if (fluxSink.requestedFromDownstream() > 0) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/AbstractProcessingStrategy.java]: is this 0 a code or a quantity? if a code, is there a constant that we can use?

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/StreamPerEventSink.java [hunk 3]
```diff
@@ -41,10 +39,18 @@ public StreamPerEventSink(ReactiveProcessor processor, Consumer<CoreEvent> event
...
     just(event)
         .doOnNext(request -> eventConsumer.accept(request))
         .transform(processor)
-        .subscribe(requestUnbounded());
+        .subscribe(null, exception::set);
+    if (exception.get() != null) {
+      if (exception.get() instanceof RuntimeException) {
+        throw (RuntimeException) exception.get();
+      } else {
+        throw new RuntimeException(exception.get());
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/StreamPerEventSink.java]: throw a MuleRuntimeException

File: pom.xml [hunk 4]
```diff
@@ -1331,6 +1331,11 @@
                 <version>${projectReactorVersion}</version>
             </dependency>
             <dependency>
+                <groupId>io.projectreactor.addons</groupId>
+                <artifactId>reactor-extra</artifactId>
```
Comment (elrodro83) [inline pom.xml]: what are you using from here?

**Comment by elrodro83** [comment]
@afelisatti Some changes here affect how errors are propagated. Can you review those?

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/ReactorProcessingStrategyTestCase.java (hunk 1 above)
Comment (dfeist) [reply core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/ReactorProcessingStrategyTestCase.java]: This was wrong before, change is unrelated to updgrade.

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/AbstractProcessingStrategy.java (hunk 2 above)
Comment (dfeist) [reply core/src/main/java/org/mule/runtime/core/internal/processor/strategy/AbstractProcessingStrategy.java]: quantity.

File: pom.xml (hunk 4 above)
Comment (dfeist) [reply pom.xml]: In all to delete all the code, that @marianogonzalez added from reactor-extra for retry. These files are removed in this PR.

File: core-tests/src/test/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicyTestCase.java [hunk 5]
```diff
@@ -183,19 +186,19 @@ public void nextProcessorExecutionFailurePropagates() throws Exception {
     reset(flowExecutionProcessor);
     when(flowExecutionProcessor.apply(any())).thenAnswer(invocation -> {
       Mono<CoreEvent> mono = from(invocation.getArgumentAt(0, Publisher.class));
-      mono.doOnNext(event -> ((BaseEventContext) event.getContext()).error(policyException)).subscribe();
+      mono.doOnNext(event -> ((BaseEventContext) event.getContext()).error(new MessagingException(event, policyException)))
```
Comment (afelisatti) [inline core-tests/src/test/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicyTestCase.java]: 😒
